docs: PEM-10477: Self-Hosted VMO GUI Troubleshooting (#10056)
spectrocloud/librarium: docs/docs-content/troubleshooting/vmo-issues.md

The following are common scenarios that you may encounter when using Virtual Machine Orchestrator (VMO).

## Scenario - VMO Loading Errors in Self-Hosted Palette

On [self-hosted Palette](../enterprise-version/install-palette/install-on-kubernetes/install-on-kubernetes.md) and
[Palette VerteX](../vertex/install-palette-vertex/install-on-kubernetes/install-on-kubernetes.md) Helm-based

> installations, the cluster's **Virtual Machines** tab or the VMO Graphical User Interface (GUI) may fail to load.

Reviewer: ```suggestion
installations, the cluster's **Virtual Machines** tab or the VMO Graphical UI (GUI) may fail to load.
```
Maybe? We don't spell out UI

### Debug Steps

To fix these issues, you must adjust two default ingress configurations. If your self-hosted environment uses an IP
address instead of a domain name, only the [Rate Limiting](#adjust-rate-limit) fix is applicable.

- [**Rate Limiting**](#adjust-rate-limit) - The `IngressRoute` resource `hubble-foreq-ingress-resource` may reference a
  rate-limit `Middleware` resource with a value that is too low, causing the VMO GUI to get stuck when loading.

- [**Content Security Policy (CSP)**](#adjust-csp) - Three Traefik `Middleware` resources contain a
  `Content-Security-Policy` header with `frame-ancestors` set to `https://*.spectrocloud.com`. This blocks the VMO GUI
  from loading when your Palette instance uses a different domain name. This issue does not apply to self-hosted
  environments that use an IP address instead of a domain name.

#### Adjust Rate Limit

1. Log in to your
   [self-hosted Palette](../enterprise-version/system-management/system-management.md#access-the-system-console) or

>    [Palette VerteX system console](../vertex/system-management/system-management.md#access-the-system-console).

Author: That's fair! I was trying to make it clear the link was specific to the system console and not the Palette/Palette VerteX landing page. But your suggestion looks cleaner.

   export KUBECONFIG=<path-to-kubeconfig>
   ```

5. Verify the current `Content-Security-Policy` value for the following `Middleware` resources.

   ```shell
   kubectl get middleware ui-csp-frame-ancestors --namespace ui-system --output yaml
   kubectl get middleware cp-csp-frame-ancestors --namespace cp-system --output yaml
   kubectl get middleware foreq-security-headers --namespace hubble-system --output yaml
   ```

   ```yaml title="Example output" hideClipboard {18}
   apiVersion: traefik.io/v1alpha1
   kind: Middleware
      metadata:
      annotations:
         meta.helm.sh/release-name: hubble
         meta.helm.sh/release-namespace: default
      creationTimestamp: "2026-04-07T15:03:54Z"
      generation: 1
      labels:
         app.kubernetes.io/managed-by: Helm
      name: ui-csp-frame-ancestors
      namespace: ui-system
      resourceVersion: "36023"
      uid: 2f87aada-b8cf-454f-8c32-38be34b060bf
   spec:
      headers:
         customResponseHeaders:
            Content-Security-Policy: frame-ancestors 'self' https://*.spectrocloud.com
   ```

6. Update each applicable `Middleware` resource by replacing `https://*.spectrocloud.com` with your root domain. Replace

>    `<rootDomain>` in the following commands with your Palette root domain.

Reviewer: ```suggestion
   `<root-domain>` in the following commands with your Palette root domain.
```
We use kebab case for placeholders

   :::tip

   Use the following command to identify your root domain.

   ```shell
   kubectl get configmap root-domain-info --namespace hubble-system --output jsonpath='{.data.apiEndpoint}'
   ```

   :::

   ```shell
   kubectl patch middleware ui-csp-frame-ancestors --namespace ui-system --type merge \
   --patch '{"spec":{"headers":{"customResponseHeaders":{"Content-Security-Policy":"frame-ancestors '\''self'\'' https://*.<rootDomain> https://<rootDomain>"}}}}'
   ```

   ```shell
   kubectl patch middleware cp-csp-frame-ancestors --namespace cp-system --type merge \
   --patch '{"spec":{"headers":{"customResponseHeaders":{"Content-Security-Policy":"frame-ancestors '\''self'\'' https://*.<rootDomain> https://<rootDomain>"}}}}'
   ```

   ```shell
   kubectl patch middleware foreq-security-headers --namespace hubble-system --type merge \
   --patch '{"spec":{"headers":{"customResponseHeaders":{"Content-Security-Policy":"frame-ancestors '\''self'\'' https://*.<rootDomain> https://<rootDomain>;"}}}}'
   ```

7. Verify the updated values. Each `Middleware` resource should display a `Content-Security-Policy` header with

>    `frame-ancestors 'self' https://*.<rootDomain> https://<rootDomain>`.

Reviewer: ```suggestion
   `frame-ancestors 'self' https://*.<root-domain> https://<root-domain>`.
```
